Modify message handling

emontnemery opened the pull request:
Modify message handling when sending status or control requests:
- When sending a request, check that the reply is matching (correct cmd, correct sequence number)
- If other messages are received, e.g. a heartbeat, pass them back to the caller

emontnemery commented on tuyaface/__init__.py:
```diff
@@ -179,22 +175,37 @@ def _process_raw_reply(device: dict, raw_reply: bytes):
     """
 
-    filtered_replies = list(filter(lambda x: x["data"] and x["data"] != 'json obj data unvalid', replies))
+    filtered_replies = list(filter(lambda x: x["data"] and x["cmd"] == tf.STATUS, replies))
```
I don't have any device which sends a valid reply to DP_QUERY.
When a DP_QUERY is sent, is the status update included in the DP_QUERY reply or is it in a STATUS message?

TradeFace replied:
It has been a while since I looked at the raw messages, not sure. I'll have to check.

TradeFace commented:
Something is missing

```
logger.debug("sending command: [%x:%s] payload: [%s]", command, tf.cmd_to_string.get(command, f'UNKNOWN'), payload)
AttributeError: module 'tuyaface.const' has no attribute 'cmd_to_string'
```

TradeFace commented:
```
2020-05-23 13:54:30,446 ERROR    (Thread-33) [tuya.tuyaface.tuyaclient] TuyaClient: Unexpected exception
Traceback (most recent call last):
  File "/home/niels/tools/tuyamqtt/tuya/tuyaface/tuyaclient.py", line 118, in run
    can_read, _, _ = select.select(rlist, [], [], HEART_BEAT_PING_TIME/2)
ValueError: file descriptor cannot be a negative integer (-1)
```
Not sure what is going on... tying to get info on you question

emontnemery commented:
weird errors, are you on right branch and so on?

TradeFace commented:
```
request 192.168.1.31 00 00 55 AA 00 00 00 01 00 00 00 **0A** 00 00 00 88 95 19 67 AB 10 9F CE 8E 7F C0 CC E2 95 41 F6 A9 59 5B DE D6 C4 2B DC B7 42 C2 8B 88 E1 FF 6B 3E 78 E2 33 86 0D 42 61 63 93 D0 C1 D4 00 A0 B8 1D 04 2A 05 20 FA 74 30 4D E5 04 06 16 05 0F 7E A1 26 DC 4A 12 89 47 91 B9 B5 C4 6E FC 2E 54 33 66 59 5B DE D6 C4 2B DC B7 42 C2 8B 88 E1 FF 6B 3E BC F9 13 5A 31 EB B7 FD A9 17 38 6E D1 20 72 AE 16 BE 43 37 31 95 C2 44 99 16 85 25 82 E0 60 74 81 A0 6E 3A 00 00 AA 55 
reply 192.168.1.31   00 00 55 AA 00 00 00 01 00 00 00 **0A** 00 00 00 2C 00 00 00 01 D6 E7 1F BA 26 5D 1B 1C 47 DB 33 1B 1B C4 3D CB E0 2A 46 00 C4 FE 92 B0 E9 A7 86 D0 82 F0 F9 4D A0 3F CF EE 00 00 AA 55 

request 192.168.1.21 00 00 55 AA 00 00 00 01 00 00 00 **0A** 00 00 00 88 8E FE 0F 29 23 44 1F 17 AC 08 48 65 25 1D F8 40 12 8D 7F 54 76 A8 94 B6 AC 4F CA B9 FE 4D 45 89 D3 5D 1A 23 E6 9B 30 B9 DE F7 CE CE D3 AC 06 D7 DE 31 F1 12 E6 2D DE 6D 3D E4 05 03 79 7A C0 27 F4 2D 83 5C C0 35 34 A3 BB 82 7D CB 26 32 79 F5 12 8D 7F 54 76 A8 94 B6 AC 4F CA B9 FE 4D 45 89 B8 D6 DF ED 9F 15 AB 4E 56 9B 4B 23 CD DE D4 5C 3C C6 87 2A 73 3C DC FF 11 E4 6D 6C 2A 00 BE 34 0D 4D F0 EB 00 00 AA 55
reply 192.168.1.21   00 00 55 AA 00 00 00 01 00 00 00 **0A** 00 00 00 2C 00 00 00 00 87 2F AE 30 33 30 5F AA 17 D9 E3 1E 3D 3F 1D 86 54 46 3B 80 D2 13 44 DA 68 CA 2A 66 ED 88 79 D0 B1 EE BE 28 00 00 AA 55 
```
It seems there are no direct status calls anymore in tuyamqtt. Only as a fallback on set_state in tuyamqtt

TradeFace commented:
Yeah, tuyamqtt @ master, tuyaface @ development

emontnemery commented:
> Yeah, tuyamqtt @ master, tuyaface @ development

Hmm, do you get these errors every time:
`AttributeError: module 'tuyaface.const' has no attribute 'cmd_to_string'`
`ValueError: file descriptor cannot be a negative integer (-1)`

emontnemery commented:
`reply 192.168.1.31   00 00 55 AA 00 00 00 01 00 00 00 **0A** 00 00 00 2C 00 00 00 01 D6 E7 1F BA 26 5D 1B 1C 47 DB 33 1B 1B C4 3D CB E0 2A 46 00 C4 FE 92 B0 E9 A7 86 D0 82 F0 F9 4D A0 3F CF EE 00 00 AA 55`
Is it a valid reply? Can you paste the decrypted version too?

TradeFace commented:
Can't you read hex? :)

https://github.com/TradeFace/tuya/wiki/Dev-Ref

emontnemery commented:
Thanks!
Updated now such that the reply to `tf.DP_QUERY` is used as status reply, unless it's `json obj data unvalid`.
Do you think it's OK?

emontnemery commented:
By the way, did you mean you get this error with wrong IP, but with right IP there is no error?
```
2020-05-23 13:54:30,446 ERROR    (Thread-33) [tuya.tuyaface.tuyaclient] TuyaClient: Unexpected exception
Traceback (most recent call last):
  File "/home/niels/tools/tuyamqtt/tuya/tuyaface/tuyaclient.py", line 118, in run
    can_read, _, _ = select.select(rlist, [], [], HEART_BEAT_PING_TIME/2)
ValueError: file descriptor cannot be a negative integer (-1)
```

emontnemery commented:
I think db7ed8a fixes the `ValueError`.

TradeFace commented:
I am not sure what happened, something was probably corrupted in my system. I've deleted the complete repo from disk and cloned it again... error is gone.

TradeFace commented:
On the PR... there is a lot to review.

The first thing what caught my eye:
```
filtered_replies = list(filter(lambda x: x["cmd"] == command, replies))
if seq is not None:
        filtered_replies = list(filter(lambda x: x["seq"] == seq, filtered_replies))
```
cmd = 7 or 13 (0D)  the devices reply with ack cmd and 08 status. status request don't have a seq.

emontnemery commented:
Yes, so the idea is to separate the reply to the command from the status update.

TradeFace commented:
When can seq `None`?

emontnemery commented:
It's just default input parameter, in case there is no need to match on seq. It's not used though and can be removed to simplify the code.

TradeFace commented:
It isn't really a problem... just caught my eye. I'll do some more reading.

TradeFace commented:
in `_process_raw_reply` you filter out any replies < 28 or not ending in XYZ. In `_status` you test for `request_reply["data"] == 'json obj data unvalid'`. Can this ever be true?

TradeFace commented:
no other remarks--

emontnemery commented:
> in `_process_raw_reply` you filter out any replies < 28 or not ending in XYZ.

The problem is that, `a.split()` returns empty slices, there's no guarantee that reading from the socket will give you the full message sent by the Tuya device and stuff like `sbytes[4:8]` doesn't error  or throw if sbytes is too short, it just returns an empty or shorter slice instead.

Try:
```diff
       for s in a.split('0x000055aa', bytealigned=True):
+        logger.info("parsing chunk '%s', len %u, %s", ''.join(format(x, '02x') for x in sbytes), len(sbytes), sbytes[9999:20000])
         sbytes = s.tobytes()
```
For every parsed package, split also returns an empty one:
```
2020-05-23 17:28:37,539 DEBUG    (Thread-2) [tuyaface] (192.168.0.179) read from socket: '000055aa000000010000000a0000002c000000018a39fdb82a4ae1b63ce95be8de91a4ac22c63d68c3d9f37d8e568d5e910afca21a9455260000aa55'
2020-05-23 17:28:37,540 INFO     (Thread-2) [tuyaface] parsing chunk '', len 0, b''
2020-05-23 17:28:37,542 INFO     (Thread-2) [tuyaface] parsing chunk '000055aa000000010000000a0000002c000000018a39fdb82a4ae1b63ce95be8de91a4ac22c63d68c3d9f37d8e568d5e910afca21a9455260000aa55', len 60, b''
```

TradeFace commented:
Okay, I didn't realize that. Beautiful code btw :+1:

TradeFace commented:
```
2020-05-23 17:42:30,937 ERROR    (Thread-16) [tuya.tuyaface.tuyaclient] TuyaClient: Unexpected exception
Traceback (most recent call last):
  File "/home/niels/tools/tuyamqtt/tuya/tuyaface/tuyaclient.py", line 150, in run
    if self._is_connection_stale():
  File "/home/niels/tools/tuyamqtt/tuya/tuyaface/tuyaclient.py", line 58, in _is_connection_stale
    self._ping()
  File "/home/niels/tools/tuyamqtt/tuya/tuyaface/tuyaclient.py", line 43, in _ping
    _send_request(self.device, tf.HEART_BEAT)
  File "/home/niels/tools/tuyamqtt/tuya/tuyaface/__init__.py", line 383, in _send_request
    logger.debug("sending command: [%x:%s] payload: [%s]", command, tf.cmd_to_string.get(command, f'UNKNOWN'), payload)
AttributeError: module 'tuyaface.const' has no attribute 'cmd_to_string'
```
This one is back... maybe it wasn't the repo

emontnemery commented:
I think you can try `print("tf.__file__")` to see where python finds the const file

TradeFace commented:
It indeed was using an other file site-packages/tuyaface/const.py. Odd. Fixed that, but found another problem.
```
<class 'dict'> #print from tuyaface:_status just before return
{'cmd': 10, 'seq': 1, 'data': '{"dps":{"1":false,"2":false,"7":0,"8":0}}', 'rc': 0} #print from tuyaclient:_status just before return
2020-05-23 19:21:20,958 ERROR    (Thread-16) [tuya.tuyaface.tuyaclient] TuyaClient: Unexpected exception
Traceback (most recent call last):
  File "/home/niels/tools/tuyamqtt/tuya/tuyaface/tuyaclient.py", line 155, in run
    result = command(*args)
  File "/home/niels/tools/tuyamqtt/tuya/tuyaface/tuyaclient.py", line 184, in _status
    data = json.loads(status_reply)
  File "/usr/lib/python3.6/json/__init__.py", line 348, in loads
    'not {!r}'.format(s.__class__.__name__))
TypeError: the JSON object must be str, bytes or bytearray, not 'dict'
```

emontnemery commented:
Oops, I'll fix that.
